### PR TITLE
fix(styled-components): avoid quadratic union distribution

### PR DESCRIPTION
**Pull request by qnighy**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57684
- [x] ~If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.~

----

After the introduction of union distribution in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57684, it started to consume a lot of memory and cause tsc to hang in a project I'm working on. It was caused by the following pattern:

```typescript
OmitU<ReactDefaultizedProps<C, P> & O, A> & Partial<PickU<P & O>, A>
```

When `P` is a union, it expands to a quadratically large union and that seems to be the cause of the memory consumption.

This PR changes `MakeAttrsOptional` so that the unions are distributed beforehand to reduce computation.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## styled-components ([<kbd>unpkg</kbd>](https://unpkg.com/browse/styled-components@latest/))
was missing the following properties:
1. StyleSheetConsumer
1. StyleSheetContext
1. version
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 021fb1c27b883df9ecf87bd3076f150ef4ae0cc5
</p>

**Review by eps1lon — CHANGES_REQUESTED** [review]
Once the Azure pipeline passes, we should get performance metrics displayed in the PR.

**Comment by qnighy** [comment]
> Once the Azure pipeline passes, we should get performance metrics displayed in the PR.

So I just need to wait for it, right?

**Comment by eps1lon** [comment]
> > Once the Azure pipeline passes, we should get performance metrics displayed in the PR.
> 
> So I just need to wait for it, right?

It failed but the error looks unrelated. I would occasionally rebase with `master` and push again to see if the related issue was fixed.

File: types/styled-components/test/index.tsx [hunk 1]
```diff
@@ -1224,6 +1224,27 @@ function unionTest() {
     <StyledReadable kind="magazine" author="Hejlsberg" />; // $ExpectError
 }
 
+function unionPerformanceTest() {
```
Comment (eps1lon) [inline types/styled-components/test/index.tsx:1227]: I don't think we need a dedicated test since if the existing perf metrics show a difference.

Let's try it without this test first and see if we see a difference. Otherwise let's separate test and fix into separate PRs (the fix based on the test once it's merged) so that we know the test is actually showing performance improvements.

Comment (qnighy) [reply types/styled-components/test/index.tsx]: Sure, I removed the test for the time being.

**Comment by eps1lon** [comment]
Hm automatic perf checks were disabled. Do you have some manual instructions how to verify that this indeed improved overall perf? Especially to check if this worked as intended for new and older versions

**Comment by eps1lon** [comment]
I'm not saying that I don't think this doesn't improve. Naively I would assume it does just by virtue of being less code. I'm more thinking about maintaining this particular change.

For starters, let's add comment to that line explaining that it has perf implications. And if it needs to change in the future, people know to manually verify perf implications since we don't have automatic checks anymore.

**Comment by qnighy** [comment]
> Hm automatic perf checks were disabled.

Seems like it finally reported the diff?

> Do you have some manual instructions how to verify that this indeed improved overall perf? Especially to check if this worked as intended for new and older versions

I adjusted the test I initially added and filed it as a separate PR so that the difference in the performance can be observed: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58176

In my local machine `npm test styled-components` takes about 33s in #58176 and about 18s in #58176 + #58131.

**Comment by eps1lon** [comment]
> Seems like it finally reported the diff?

Yeah, I asked in the DT discord to run them manually. Though I have no idea if these changes are statistically significant. Let's get https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58176 in, rebase and then I ask for a perf run again.

**Comment by kolyasapphire** [comment]
Thank you very much for looking into this! I think we are affected by this issue.

We are currently still on `"@types/styled-components": "5.1.17"`. Our automated workflow runs to bump the version fail on `Checking validity of types` stage of Next.js build with:

```
FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
```

Same happens locally.

Using `qnighy:styled-components-quadratic` fork solved the issue. I saved both the fork and the original repo and used Yarn's `link:` feature to try them.

**Comment by eps1lon** [comment]
@kolyasapphire Thanks for the confirmation. I'd still like to confirm that we can repro the current bad behavior with tests added in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58176. If not, then we just add a comment.

**Comment by qnighy** [comment]
@eps1lon now that https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58176 has been merged, I rebased this PR on top of that. Could you take a look at it again?

**Comment by qnighy** [comment]
Build failing due to https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58270, I guess?

**Comment by eps1lon** [comment]
I've asked for another perf run. But even if it comes back not conclusive I'd approve it to get a merge.

**Review by eps1lon — APPROVED** [review]

**Comment by kolyasapphire** [comment]
Thanks a lot! Upgraded the package, builds are working again
